feat (static table): implement a read-only table struct loaded from metadata

a-agmon opened the pull request:
Implements #246 
The PR adds a `StaticTable` struct that encapsulates `Table` and enables loading a table from meta-data, regardless of a catalog but provides only read-only access to meta-data and scan. 
This aims to provide similar functionality to the StaticTable in pyIceberg and its main value is to make table metadata and scan accessible for various operations using only the metadata.json, in a typesafe way.

Added:
The PR also adds an `into_table()` method that returns the `StaticTable` as a `Table`. In addition, it adds a boolean readonly flag to `Table` which helps to distinguish between a readonly Table built using a StaticTable and a Table loaded from Catalog.

liurenjie1024 commented on crates/iceberg/src/static_table.rs:
````diff
@@ -0,0 +1,146 @@
+//! `StaticTable` is a read-only table struct that can be created from a metadata file or from `TableMetaData` without a catalog.
+//! It can only be used to read metadata and for table scan.
+//! For example:
+//! ```rust, compile_fail
````
Why this compile failed? I think the compilation can pass by adding some `using` statement?

a-agmon replied:
I added the `compile_fail` flag because the example involves async code that requires runtime, and I didn't want to mess up the code example. I will add it.

liurenjie1024 commented on crates/iceberg/src/static_table.rs:
```diff
@@ -0,0 +1,146 @@
+impl StaticTable {
+    /// creates a static table from a given `TableMetadata` and `FileIO`
+    pub async fn from_metadata(metadata: TableMetadata, file_io: FileIO) -> anyhow::Result<Self> {
+        let static_identifier = TableIdent::from_strs(["static_ns", "static_table"])?;
```
It's better to allow user to pass `table_identifier`?

a-agmon replied:
No problem, shall I wrap it in an Option to make it optional?

liurenjie1024 replied:
I prefer not, user can choose their default name and it's not huge effort.

liurenjie1024 commented on crates/iceberg/src/static_table.rs:
```diff
@@ -0,0 +1,146 @@
+        Ok(Self(table))
+    }
+    /// creates a static table directly from metadata file and `FileIO`
+    pub async fn from_metadata_file(
```
Ditto.

liurenjie1024 commented on crates/iceberg/src/static_table.rs:
```diff
@@ -0,0 +1,146 @@
+use futures::AsyncReadExt;
+
+/// `StaticTable` is a read-only table struct that can be created from a metadata file or from `TableMetaData` without a catalog.
+pub struct StaticTable(Table);
```
How about moving this to `table` module given they are closely related?

a-agmon replied:
No problem.

liurenjie1024 commented on crates/iceberg/src/static_table.rs:
```diff
@@ -0,0 +1,146 @@
+        table_metadata_location: String,
+    }
+
+    impl TableTestFixture {
```
We don't need this as we already have valid table metadata: crates/iceberg/testdata/table_metadata/TableMetadataV2Valid.json

The reason we need the template metadata file because we need to test load manifest api, and it's not necessary here.

a-agmon replied:
Will fix this

liurenjie1024 reviewed (commented):
Hi, @a-agmon This pr looks great! Some small improvements would make it better.

a-agmon commented:
Hi @liurenjie1024 ,
Thanks much for the comments!
I have addressed all issues. For the doctest, I added the tokio-test crate as a dev dependency in order to have an async runtime there for the async code. This seems like the right [approach](https://stackoverflow.com/a/64597248/22560711).
 It only test compile as I think its a good idea to show an example using S3 even though it can't run. At any rate this flow is unit tested. 

Hope this is better

liurenjie1024 commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,98 @@ impl Table {
+pub struct StaticTable(Table);
+
+impl StaticTable {
+    /// creates a static table from a given `TableMetadata` and `FileIO`
```
```suggestion
    /// Creates a static table from a given `TableMetadata` and `FileIO`
```

liurenjie1024 commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,98 @@ impl Table {
+            .build();
+        Ok(Self(table))
+    }
+    /// creates a static table directly from metadata file and `FileIO`
```
```suggestion
    /// Creates a static table directly from metadata file and `FileIO`
```

liurenjie1024 commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,98 @@ impl Table {
+        Self::from_metadata(table_metadata, table_ident, file_io).await
+    }
+
+    /// create a TableScanBuilder for the static table
```
```suggestion
    /// Create a TableScanBuilder for the static table.
```

liurenjie1024 commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,98 @@ impl Table {
+        self.0.scan()
+    }
+
+    /// get a TableMetadataRef for the static table
```
```suggestion
    /// Get  TableMetadataRef for the static table
```

liurenjie1024 reviewed (approved):
Hi, @a-agmon This pr looks great! We can ship it after fixing some small typos.

Xuanwo commented on crates/iceberg/src/table.rs:
````diff
@@ -63,3 +64,99 @@ impl Table {
+/// .snapshot_id();
+/// # });
+/// ```
+pub struct StaticTable(Table);
````
`static` has special meanings in Rust. Have we considered using `ReadOnlyTable` for greater expressiveness?

a-agmon replied:
Thank @Xuanwo 
I think the idea was to be consistent with the [Python API](https://py.iceberg.apache.org/reference/pyiceberg/table/#pyiceberg.table.StaticTable) that uses StaticTable for this functionality. I'm also OK with ReadOnlyTable

Fokko replied:
It is also called StaticTable in Java: https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/StaticTableOperations.java

Xuanwo commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,99 @@ impl Table {
+        metadata: TableMetadata,
+        table_ident: TableIdent,
+        file_io: FileIO,
+    ) -> anyhow::Result<Self> {
```
How about using `iceberg::Result` here?

a-agmon replied:
Done. Thanks

Xuanwo commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,99 @@ impl Table {
+        metadata_file_path: &str,
+        table_ident: TableIdent,
+        file_io: FileIO,
+    ) -> anyhow::Result<Self> {
```
The same.

a-agmon replied:
Done

Xuanwo commented on crates/iceberg/src/table.rs:
```diff
@@ -63,3 +64,99 @@ impl Table {
+/// # use iceberg::io::FileIO;
+/// # use iceberg::table::StaticTable;
+/// # use iceberg::TableIdent;
+/// # tokio_test::block_on(async {
```
We can avoid introducing `tokio_test` by writing an async function:

```rust
#async fn test() {
    let metadata_file_location = "s3://bucket_name/path/to/metadata.json";
    ...
#}
```

a-agmon replied:
@Xuanwo - Great idea. Thanks.
Will remove the dependency

a-agmon commented:
Thank you @liurenjie1024 and @Xuanwo for the comments!
Besides one line break that I can't get the formatter to accept, I believe I have addressed all comments. 
I can also add something like this to this PR if there is an agreement

```rust
    /// Consumes the `StaticTable` and return it as a `Table`
    /// Please use this method carefully as the Table it returns remains detached from a catalog
    /// and can't be used to perform modifications on the table. 
    pub fn into_table(self) -> Table {
        self.0
    }
```

Xuanwo commented:
> Besides one line break that I can't get the formatter to accept, I believe I have addressed all comments.

Looks fine to me. Waiting for @liurenjie1024's comment.

a-agmon commented:
Hi @liurenjie1024 and @Xuanwo 
PR is now updated with comments from our discussion [here](https://github.com/apache/iceberg-rust/issues/246#issuecomment-1998709854)
It adds a readonly flag to `Table`, and sets it to false by default. 
On `StaticTable` The readonly flag is set to true, and can return a readonly `Table` using `into_table()`

Thanks for the comments, this API will be very useful for my use case, in which I can safely pass along a StaticTable that will be treated as readonly.

Fokko reviewed (approved):
This looks good @a-agmon 🙌 Thanks for the review @Xuanwo and @liurenjie1024
